pkg/servicequota/servicequota.go

```diff
@@ -0,0 +1,89 @@
+// Package servicequota check if you have enough resources in aws before cluster creation starts
+package servicequota
+
+import (
+	"encoding/json"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/service/servicequotas"
+	"github.com/hjson/hjson-go"
+	"github.com/oslokommune/okctl/pkg/api/okctl.io/v1alpha1"
+	"github.com/tidwall/gjson"
+)
+
+// Usage defines what we need to know about a service quota
+type Usage interface {
+	Count() (int, error)
+	Quota() (int, error)
+	Required() int
+	CheckAvailability() error
+	Description() string
+}
+
+// CheckQuotas check if we have enough resources for given cloud provider
+func CheckQuotas(provider v1alpha1.CloudProvider) error {
+	vpcs := vpcusage{}
+	vpcs.CloudProvider = provider
+
+	err := vpcs.CheckAvailability()
+	if err != nil {
+		return err
+	}
+
+	eips := eipusage{}
+	eips.CloudProvider = provider
+
+	err = eips.CheckAvailability()
+	if err != nil {
+		return err
+	}
+
+	igws := igwusage{}
+	igws.CloudProvider = provider
+
+	err = igws.CheckAvailability()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getValueFromHjson(humanjson string, path string) string {
+	var result map[string]interface{}
+
+	err := hjson.Unmarshal([]byte(humanjson), &result)
+	if err != nil {
+		return ""
+	}
+
+	marshalled, err := json.Marshal(result)
+	if err != nil {
+		return ""
+	}
+
+	return gjson.Get(string(marshalled), path).String()
+}
+
+func getStringMapOf(humanJSON string) map[string]interface{} {
+	var result map[string]interface{}
+
+	err := hjson.Unmarshal([]byte(humanJSON), &result)
+	if err != nil {
+		return nil
+	}
+
+	return result
+}
+
+func getLengthOf(result map[string]interface{}, path string) (int, error) {
+	if result[path] == nil {
+		return 0, nil
+	}
+
+	return len(result[path].([]interface{})), nil
+}
+
+func getQuotaNumericValue(quota *servicequotas.GetServiceQuotaOutput) (int, error) {
+	return strconv.Atoi(getValueFromHjson(quota.String(), "Quota.Value"))
```

Comment:
Rather than fetching the value this way, we can derefence the value in the returned output: https://docs.aws.amazon.com/sdk-for-go/api/service/servicequotas/#ServiceQuota.

e.g.:

```go
output := &ServiceQuota{}
value := *output.Value
```

Reply:
Could you provide a more complete example?  I don't get what you are saying here at all.

Reply:
Absolutely! Does this snippet of code demonstrate it better: https://github.com/oslokommune/okctl/blob/master/pkg/route53/route53.go#L38-L58?

Basically you don't need to use the `.String()` function on the output and parse that output. You can rather dereference the field you are interested in. So, for the `ServiceQuota` struct, we have this form:

```go
type ServiceQuota struct {

    // Specifies if the quota value can be increased.
    Adjustable *bool `type:"boolean"`

    // Specifies the ErrorCode and ErrorMessage when success isn't achieved.
    ErrorReason *ErrorReason `type:"structure"`

    // Specifies if the quota is global.
    GlobalQuota *bool `type:"boolean"`

    // Identifies the unit and value of how time is measured.
    Period *QuotaPeriod `type:"structure"`

    // The Amazon Resource Name (ARN) of the service quota.
    QuotaArn *string `type:"string"`

    // The code identifier for the service quota specified.
    QuotaCode *string `min:"1" type:"string"`

    // The name identifier of the service quota.
    QuotaName *string `type:"string"`

    // Specifies the service that you want to use.
    ServiceCode *string `min:"1" type:"string"`

    // The name of the AWS service specified in the increase request.
    ServiceName *string `type:"string"`

    // The unit of measurement for the value of the service quota.
    Unit *string `type:"string"`

    // Specifies the details about the measurement.
    UsageMetric *MetricInfo `type:"structure"`

    // The value of service quota.
    Value *float64 `type:"double"`
    // contains filtered or unexported fields
}
```

If you have an instance of this type above, which you have in this receiver: `quota *servicequotas.GetServiceQuotaOutput`, you can simply dereference `Value` for example using `*quota.Quota.Value`

Reply:
Ah ... this makes a lot of sense, so now i can get rid of the humanJson parsing bit, thanks! :)